[SYCL] Update clang driver AOT design to support optional features. (#4202)
intel/llvm: sycl/doc/OptionalDeviceFeatures.md

#### Device configuration file

The configuration file uses a simple YAML format where each top-level key is
a name of a device architecture. There are sub-keys under each device for

> the supported aspects, sub-group sizes and AOT compiler ID.  For example:

Reviewer: ```suggestion
the supported aspects and sub-group sizes.  For example:
```

*Command line arguments*:

- `-target=<target>` target device architecture to filter for
- `-device-config-file=<path>` path to the device configuration file

*Output*:

- the input file table filtered as needed

In more details, the tool performs the following actions:

1) Checks if the input file table contains "Properties" column. If not, copies
   the input file table to output and exits without error.
1) Reads in the device configuration file and finds some entry `E` corresponding
   to the architecture given on the command line. If there is no such entry -
   reports an error and exits.
1) For each row in the input file table:

   - loads the properties file from the "Properties" column
   - checks if there is the `SYCL/device-requirements` property
   - if no, copies current row to the output file table and goes to the next
   - if yes, checks if all the requirements listed in the property are supported
     by the target architecture as specified by entry `E` in the device
     configuration file
       - if yes, copies current row to the output file table and goes to the
         next
       - otherwise skips this row

##### Configuration file location and driver option

> A default device configuration file is supplied as a part of oneAPI SDK. It is
> located in the TBD directory. Users may override the default using the
> `-fsycl-device-config-file=<path>` option.

Reviewer 2: ```suggestion
A default device configuration file is present in DPC++ build. Users may override it using the
`-fsycl-device-config-file=<path>` compiler command line option.
Exact location of the file and final name of the compiler option is TBD.
```

##### AOT target identification

There are several user-visible places in the SDK where SYCL device target
architectures need to be identified:

- `-fsycl-targets` option
- a device configuration file entry
- `-target` option of the `sycl-aspec-filter` tool
- a SYCL aspect enum identifier (we expect to add a new SYCL aspect for each
  device target architecture)

In all such places architecture naming should be the same. In some cases aliases
are allowed. Below is a list of target architectures supported by DPC++:

| target/alias(es) | description |
|-|-|
| intel_gpu    | Generic Intel graphics architecture |
| intel_gpu_tgl, intel_gpu_12_0 | Intel Tiger Lake (11th generation Core) integrated graphics architecture |
| ptx64  | Generic 64-bit PTX target architecture |
| spir64 | Generic 64-bit SPIR-V target |
| x86_64 | Generic 64-bit x86 architecture |

TODO: Provide full list of AOT targets supported by the identification
mechanism.

Example of clang compilation invocation with 2 AOT targets and generic SPIR-V:
```

> clang++ -fsycl -fsycl-targets=spir64,intel_gpu_12_0,ptx64 ...

Reviewer 3: A belated note: this example is not correct, since the `-fsycl-targets=` option doesn't support Intel Graphics platform specification (that is done via `-Xsycl-target-backend "-device ..."` mechanism of option passthrough to `ocloc`). The closest a compiler user would be able to get to this would be:
```
clang++ -fsycl -fsycl-targets=spir64,spir64_gen,nvptx64-nvidia-cuda -Xsycl-target-backend=spir64_gen "-device gen12" ...
```
However, I realize that some interface changes may be planned in this regard that are not covered by docs for option device features. So what would be the correct way to reformulate this bit, given the planned compiler architecture changes?

> fact, the DPC++ driver supports a command line option which allows the user

Reviewer: I suggest you remove the following sentences from the first paragraph in this section because the device names are discussed in more depth in the new section you added.

> We expect to define a set of device architecture names that are used consistently in many places (in this
configuration file, in the names of device-specific aspects, as parameters for the `-fsycl-targets` command line option, etc.)  However, we have not yet agreed on these architecture names.

After removing this, the paragraph will state simply:

> The configuration file uses a simple YAML format where each top-level key is a name of a device architecture.  There are sub-keys under each device for the supported aspects, sub-group sizes and AOT compiler ID.  For example:

Author: I'll try to reformulate. I think a note that the design requires to have all architecture names consistent everywhere in the compiler is still useful.

Reviewer: I agree that it's important to say this, but you already say it in the section "AOT target identification".  There's no need to duplicate it also in the introduction.

Author: ok, I see. Will do.

to select an alternate configuration file.

**TODO**: 
* The names of the devices in the configuration file.

> * Location of the default device configuration file

Reviewer: I suggest you remove the TODO "The names of the devices in the configuration file" because @bader has already suggested you add a TODO about this in the section "AOT target identification".

Likewise, I suggest you move the TODO "Location of the default device configuration file" below to the section titled "Configuration file location and driver option".

modules - it outputs a file table as a result. Each row in the table corresponds
to an individual output module, and each element of a row is a name of a file
containing necessary information about the module, such as the code itself, its
properties.

At the action graph building stage for each requested AOT compilation target -
SPIR-V-based (such as Gen targets) and/or non-SPIR-V-based (such as PTX) - the
driver adds an `aspect-filter` action which filters out input file table rows
with device code modules using features unsupported on current target. Then the
output table goes as input into the AOT stage, and the prior filtering
guarantees that the AOT compiler will not encounter device code it can't
compile. In the extreme case when all device code
modules use unsupported aspects, the input file table will be empty. The picture
below illustrates the action graph built by the clang driver along with file
lists and tables generated and consumed by various nodes of the graph. The
example set of targets used for the illustration is 4 targets
- spir64 (runtime JITted SPIR-V)
- AOT targets
    - non-SPIR-V based
        - nvptx (PTX)
    - SPIR-V based
        - gen_11 (Intel Gen)

>         - x86_64_avx512 (AVX512)

Reviewer: It would be good to align the target names here and the names in the SVG file with the names you propose below in "AOT target identification".  For example, there is no target named "nvptx", so this should probably list "ptx64" instead.  Also, there is no entry "gen_11" in the table below.  Should that be added to the table?  Or, maybe this example should use "intel_gpu_9_0" instead?  Note that these names also show up in several places in the SVG diagram, so those should be updated too.

Also, update the example configuration file above in the section titled "Device configuration file" to use device names that you propose.

Author: right, good catch

<br>
<br>

> ![Device SPIRV translation and AOT compilation](images/DeviceLinkAOTAndWrap.svg)

Reviewer: I think it might be helpful to change the wording in the boxes in this SVG file that are currently labeled "Aspect filter" to show the real tool name and its command line.  Therefore, a typical box would have text like:

```
sycl-feature-filter -target=ptx64
```

This will help connect the diagram to the tool description below.

Author: > I think it might be helpful to change the wording in the boxes in this SVG file that are currently labeled "Aspect filter" to show the real tool name and its command line.

OK

> sycl-feature-filter

the tool really filters based on aspects, so I'd stay with sycl-aspect-filter, it also gives a connection to the spec

> ##### Aspect filter tool

Reviewer: I suggest renaming this tool to `sycl-feature-filter` because it will need to filter out lines in the file table based on features that a device supports, and not all of these features correspond to aspects.  For example, the `reqd_sub_group_size` feature does not have a corresponding aspect.

Author: The spec says that  `reqd_sub_group_size` and `reqd_work_group_size` are the only features not mapping to aspect. Also 'aspect' is more specific than 'feature', as it is defined by the SYCL spec. I would say sub/work group sizes are special cases of aspects, so aspect in the name of the tool seems more informative to me.

This tool transforms an input file table by removing rows with device code files
that use features unsupported for the target architecture given as tool's
argument.

*Name*:
- `sycl-aspect-filter`, located next to other tools like `file-table-tform`

*Input*:
- file table, normally coming out of `sycl-post-link` or `file-table-tform`
  tools

*Command line arguments*
- `-target=<target>` target device architecture to filter for
- `-device-config-file=<path>` path to the device configuration file

> *Output*
> - the input file table filtered as needed

Reviewer 4: ```suggestion
*Output*:

- the input file table filtered as needed
```

In more details, the tool performs the following actions:
1) Checks if the input file table contains "Properties" column. If not, copies
   the input file table to output and exits without error.
1) Reads in the device configuration file and finds some entry `E` corresponding
   to the architecture given on the command line. If there is no such entry -
   reports and error and exits.
1) For each row in the input file table:
   - loads the properties file from the "Properties" column
   - checks if there is the `SYCL/image-requirements` property
   - if no, copies current row to the output file table and goes to the next
   - if yes, checks if all the requirements listed in the property are supported
     by the target architecture as specified by entry `E` in the device configuration file
       - if yes, copies current row to the output file table and goes to the
         next
       - otherwise skips this row

##### Configuration file location and driver option

> A default device configuration file is supplied as a part of OneAPI SDK. It is

Reviewer 4: ```suggestion
A default device configuration file is supplied as a part of oneAPI SDK. It is
```

> located in the TBD directory. Users may override the defalt using the
> ```
> -fsycl-device-config-file=<path>
> ```
> option.

Reviewer 4: ```suggestion
located in the TBD directory. Users may override the default using the
`-fsycl-device-config-file=<path>` option.
```

##### AOT target identification

There are several user-visible places in the SDK where SYCL device target
architectures need to be identified:
- `-fsycl-targets` option
- a device configuration file entry
- `-target` option of the `sycl-aspec-filter` tool

> - a SYCL aspect enum identifier

Reviewer: ```suggestion
- a SYCL aspect enum identifier (we expect to add a new SYCL aspect for each device target architecture)
```

In all such places architecture naming should be the same. In some cases aliases
are allowed. Below is a list of target architectures supported by DPC++:

| target/alias(es) | description |
|-|-|
| fpga   | Generic Intel FPGA accelerator architecture |

> | intel_gpu    | Generic Intel graphics architecture |

Reviewer: I'm not sure about these first two entries because they don't identify a specific target device.  Will it be legal to AOT-compile with `-fsycl-targets=fpga` or `-fsycl-targets=intel_gpu`?  Don't you need to name a specific target device instead?

Author: Yes, they should be legal and reflect some generic device architecture, such that code compiled with this target architecture will run on any device of this kind. In fact, intel_gpu might mean generating code for every arch supported by `ocloc`.  BTW, `fpga` AOT target already exists, so we need to support it. See e.g. https://github.com/intel/llvm-test-suite/blob/intel/SYCL/AOT/accelerator.cpp

| intel_gpu_skl, intel_gpu_9_0 | Intel Skylake (6th generation Core) integrated graphics architecture |
| intel_gpu_kbl, intel_gpu_9_1 | Intel Kaby Lake (7th generation Core) integrated graphics architecture |
| ptx64  | Generic 64-bit PTX target architecture |
| spir64 | Generic 64-bit SPIR-V target |
| x86_64 | Generic 64-bit x86 architecture |

> More targets TBD

Reviewer 4: ```suggestion
TODO: add more targets
```
